geo/geos: change loading pattern of the GEOS dynamic library

otan opened the pull request:
Change the GEOS library to be lazily loaded, with an explicit
instruction to load when the server initializes. Furthermore, expose the
functionality to return a detailed error message (with path names, etc)
for logs and a bare bones message for CRDB users.

The lazy loading makes it easier for tests to not require loading it up
themselves each time, whilst the server initializing allows us to display
a full error message in the logs whereas people performing geospatial
operations without GEOS properly set up do not see internal server set
up info.

Also fixed the dlsym/dlerr access pattern to check whether the dlsym'd
object is NULL before returning dlerr. This errors on some systems with
`_nss_cache_cycle_prevention_function`.

Resolves https://github.com/cockroachdb/cockroach/issues/47332

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/47348)
<!-- Reviewable:end -->

petermattis reviewed (commented):
General approach <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47348#-:-M4_IvBS10ujsWsOxLvk:b-w1typl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/geo/geos/geos_unix.cc, line 129 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47348#-M4_H8051ck9nd07ojcH:-M4_H8051ck9nd07ojcI:b-3oqgu1) ([raw file](https://github.com/cockroachdb/cockroach/blob/9fb5679b8cc698ef069d0dedb1080d213ce63573/pkg/geo/geos/geos_unix.cc#L129)):*
> ```cpp
>       return dlerror();
>     }
>     return NULL;
> ```

`s/NULL/nullptr/g`

`nullptr` is the modern C++ term. `NULL` is C. Almost, but not quite identical.
___
*[pkg/geo/geos/geos_unix.go, line 40 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47348#-M4_IaRSDMljITBPGk4H:-M4_IaRSDMljITBPGk4I:bs3wkhh) ([raw file](https://github.com/cockroachdb/cockroach/blob/9fb5679b8cc698ef069d0dedb1080d213ce63573/pkg/geo/geos/geos_unix.go#L40)):*
> ```Go
> // This should only be touched by "fetchGEOSOrError".
> type crGEOSOnce struct {
> 	crGEOS *C.CR_GEOS
> ```

The `CR_` prefix makes sense for the C types. Not sure if we need the `cr` prefix on the Go types and variables. 
___
*[pkg/geo/geos/geos_unix.go, line 45 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47348#-M4_HSdi4njXsrxNe2pv:-M4_HSdi4njXsrxNe2pw:b-a18zl6) ([raw file](https://github.com/cockroachdb/cockroach/blob/9fb5679b8cc698ef069d0dedb1080d213ce63573/pkg/geo/geos/geos_unix.go#L45)):*
> ```Go
> }
> 
> var crGEOS crGEOSOnce
> ```

You can actually combine the type and var into one clause:

```
var crGEOSOnce struct {
  crGEOS *C.CR_GEOS
  err error
  once sync.Once
}
```

Functionally the same. This is usually done when you don't need to explicitly refer to the type, as is the case here.
___
*[pkg/server/server.go, line 1045 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47348#-M4_JxXoCov8LNLai1Og:-M4_JxXoCov8LNLai1Oh:b4xfbyx) ([raw file](https://github.com/cockroachdb/cockroach/blob/9fb5679b8cc698ef069d0dedb1080d213ce63573/pkg/server/server.go#L1045)):*
> ```Go
> 	)
> 
> 	if _, err := geos.FetchGEOSOrError(geos.FetchGEOSOrErrorDisplayPrivate); err != nil {
> ```

A bit of repetition in the term `FetchGEOSOrError` in this line. Using the term `geos` in both the package, function, and constant stutters. Perhaps this should be something like:

```
if err := geos.EnsureInit()
```

Inside the `geos` package you'd have an `ensureInit() (*C.CR_GEOS, error)` method which `EnsureInit() error` would call.
___
*[pkg/server/server.go, line 1047 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47348#-M4_JlGgCs0Ln3tXsOFf:-M4_JlGgCs0Ln3tXsOFg:b6vw9no) ([raw file](https://github.com/cockroachdb/cockroach/blob/9fb5679b8cc698ef069d0dedb1080d213ce63573/pkg/server/server.go#L1047)):*
> ```Go
> 	if _, err := geos.FetchGEOSOrError(geos.FetchGEOSOrErrorDisplayPrivate); err != nil {
> 		log.Infof(ctx, "could not initialize GEOS - geospatial functions may not be available: %v", err)
> 	}
> ```

I'm not sure if this is the final resting place for this initialization, but it seems fine for now.


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47348)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @petermattis)

___
*[pkg/geo/geos/geos_unix.cc, line 129 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47348#-M4_H8051ck9nd07ojcH:-M4_MTBa9hUvqJiK0-fK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9fb5679b8cc698ef069d0dedb1080d213ce63573/pkg/geo/geos/geos_unix.cc#L129)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`s/NULL/nullptr/g`

`nullptr` is the modern C++ term. `NULL` is C. Almost, but not quite identical.
</blockquote></details>

Done.
___
*[pkg/geo/geos/geos_unix.go, line 40 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47348#-M4_IaRSDMljITBPGk4H:-M4_MxmM0MkqVBpFHdEl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9fb5679b8cc698ef069d0dedb1080d213ce63573/pkg/geo/geos/geos_unix.go#L40)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

The `CR_` prefix makes sense for the C types. Not sure if we need the `cr` prefix on the Go types and variables. 
</blockquote></details>

Done.
___
*[pkg/geo/geos/geos_unix.go, line 45 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47348#-M4_HSdi4njXsrxNe2pv:-M4_MxyD64t1gIBHPWTa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9fb5679b8cc698ef069d0dedb1080d213ce63573/pkg/geo/geos/geos_unix.go#L45)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

You can actually combine the type and var into one clause:

```
var crGEOSOnce struct {
  crGEOS *C.CR_GEOS
  err error
  once sync.Once
}
```

Functionally the same. This is usually done when you don't need to explicitly refer to the type, as is the case here.
</blockquote></details>

Done.
___
*[pkg/server/server.go, line 1045 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47348#-M4_JxXoCov8LNLai1Og:-M4_O1La-kACUUAV8sWc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9fb5679b8cc698ef069d0dedb1080d213ce63573/pkg/server/server.go#L1045)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

A bit of repetition in the term `FetchGEOSOrError` in this line. Using the term `geos` in both the package, function, and constant stutters. Perhaps this should be something like:

```
if err := geos.EnsureInit()
```

Inside the `geos` package you'd have an `ensureInit() (*C.CR_GEOS, error)` method which `EnsureInit() error` would call.
</blockquote></details>

Done.
___
*[pkg/server/server.go, line 1047 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47348#-M4_JlGgCs0Ln3tXsOFf:-M4_O14UAFGbISdnEFMA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9fb5679b8cc698ef069d0dedb1080d213ce63573/pkg/server/server.go#L1047)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I'm not sure if this is the final resting place for this initialization, but it seems fine for now.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

petermattis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> module remaining nit about `NULL`.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47348#-:-M4_jhOOEumqqpzVmqUv:b6rehv0)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/geo/geos/geos_unix.cc, line 129 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47348#-M4_H8051ck9nd07ojcH:-M4_jWQ658UqeYPXd5Av:bz70kvg) ([raw file](https://github.com/cockroachdb/cockroach/blob/9fb5679b8cc698ef069d0dedb1080d213ce63573/pkg/geo/geos/geos_unix.cc#L129)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Heh, my recommendation was the opposite: use `nullptr`, not `NULL`. 


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/geo/geos/geos_unix.cc, line 129 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47348#-M4_H8051ck9nd07ojcH:-M4_kyC3CAiZNVqwrfZ8:b-48rflb) ([raw file](https://github.com/cockroachdb/cockroach/blob/9fb5679b8cc698ef069d0dedb1080d213ce63573/pkg/geo/geos/geos_unix.cc#L129)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Heh, my recommendation was the opposite: use `nullptr`, not `NULL`. 
</blockquote></details>

o wooooooops


<!-- Sent from Reviewable.io -->

otan committed:
geo/geos: change loading pattern of the GEOS dynamic library

Change the GEOS library to be lazily loaded, with an explicit
instruction to load when the server initializes. Furthermore, expose the
functionality to return a detailed error message (with path names, etc)
for logs and a bare bones message for CRDB users.

The lazy loading makes it easier for tests to not require loading it up
themselves each time, whilst the server initializing allows us to display
a full error message in the logs whereas people performing geospatial
operations without GEOS properly set up do not see internal server set
up info.

Also fixed the dlsym/dlerr access pattern to check whether the dlsym'd
object is NULL before returning dlerr. This errors on some systems with
`_nss_cache_cycle_prevention_function`.

Release note: None

otan commented:
tftr! bors r=petermattis

otan commented:
bors r=petermattis

bors needs it on it's own line apparently.
